src/main/scala/io/iohk/ethereum/consensus/blocks/BlockGenerator.scala

```diff
@@ -0,0 +1,62 @@
+package io.iohk.ethereum.consensus.blocks
+
+import akka.util.ByteString
+import io.iohk.ethereum.domain.{Address, Block, SignedTransaction}
+import io.iohk.ethereum.ledger.BlockPreparationError
+
+/**
+ * We use a `BlockGenerator` to create the next block.
+ * In a PoW setting, this is what a miner typically does.
+ * In general, a [[BlockGenerator]] depends on and is provided by the
+ * [[io.iohk.ethereum.consensus.Consensus Consensus]].
+ *
+ * @note This is generally a stateful object.
+ *
+ * @see [[io.iohk.ethereum.consensus.Consensus#blockGenerator Consensus#blockGenerator()]],
+ *      [[io.iohk.ethereum.ledger.BlockPreparator BlockPreparator]]
+ *
+ */
+trait BlockGenerator {
+  /**
+   * The type of consensus-specific data used in the block generation process.
+   * For example, under [[io.iohk.ethereum.consensus.ethash.EthashConsensus EthashConsensus]],
+   * this represents the [[io.iohk.ethereum.network.p2p.messages.PV62.BlockBody#uncleNodesList ommers]].
+   */
+  type X
+
+  /** An empty `X` */
+  def emptyX: X
+
+  // FIXME This is currently used only for Ethash
+  def getPrepared(powHeaderHash: ByteString): Option[PendingBlock]
+
+  /**
+   * This function returns the block currently being mined block with highest timestamp
+   */
+  def getPendingBlock: Option[PendingBlock]
+
+  def getPendingBlockAndState: Option[PendingBlockAndState]
+
+  def blockTimestampProvider: BlockTimestampProvider
```

Comment:
why does it need to be a part of BlockGenerator interface? It's not used anywhere other than specific implementations

Reply:
Indeed. I [moved it](https://github.com/input-output-hk/mantis/pull/410/commits/e218b514624e7bf256a40776e837e1c971146dd9) in `TestBlockGenerator`. 

Note: although technically not used from a `TestBlockGenerator` either, we do have a method there that provides a new `BlockTimestampProvider` that is used from a test. So it is OK, since we introduce the concept of this provider there.